infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -249,13 +239,37 @@ private void createSessions() {
                 incomingSessions.size(), sessions);
     }
 
+
     private void logNewlyAddedSessionInfo(Set<LogReplicationSession> newlyAddedSessions) {
         log.info("========= HashCode -> Session mapping for newly added session: =========");
         for (LogReplicationSession session : newlyAddedSessions) {
             log.info("HashCode: {}, Session: {}", session.hashCode(), session);
         }
     }
 
+    /**
+     * Build a session and add to incoming/outgoing session appropriately.
+     */
+    private void sessionBuilder(String sourceClusterId, String sinkClusterId,
+                                Set<LogReplicationSession> sessionsToAdd, TxnContext txn,
+                                Map<String, ClusterDescriptor> connectionEndpoints, boolean isIncoming) {
+        LogReplicationSession session = LogReplicationSession.newBuilder()
+                .setSourceClusterId(sourceClusterId)
+                .setSinkClusterId(sinkClusterId)
+                .setSubscriber(getDefaultSubscriber())
+                .build();
+
+        if (!sessions.contains(session)) {
+            sessionsToAdd.add(session);
+            if (connectionEndpoints.containsKey(sourceClusterId) || connectionEndpoints.containsKey(sinkClusterId)) {
```

Comment:
I don't understand this, why would it be an outgoing session if the source "or" the sink is present? an outgoing session would be if this node is the source, what does connectionEndpoints represent?

Reply:
connectionEndpoints has the remote clusters to which the local cluster will initiate connection to.
Like currently in FULL_TABLE usecase, the connectionEndpoints will have the SINKs when local cluster is the SOURCE, and it will be empty when local cluster is the SINK.

So if either source or sink (this is a little bit extending into the possibility of having a sink initiate the connection) is present in the connectionEndpoint, it would mean that the local cluster would be initiating the connection. Hence outgoing Session

Reply:
Just trying to understand this - we are checking both source and sink against `connectionEndpoints` so even if sink is initiating the connection why does it need to go into outgoing sessions?
nit: would it help if we call it `remoteClusterEndpoints` ?

Reply:
Sure, I'll rename the connectionEndpoints to remoteCLusterEndpoints.

I am thinking that outgoingSession will be the ones who initiate connection..because that is the one who is going to setup a stream on which replication will happen. Its just a grouping mechanism I thought might be good to have it consistent across the code. This is what I am thinking for the connectionModel PR. When I rebase that one on top of this, things will look a bit different.

But if its confusing, I'll remove checking the sink. Strictly speaking, checking remoteClusterEndpoints is not needed in this PR. Since I know whats coming next, I have used it here.
Please let me know.

Reply:
have removed this bit to avoid confusion